lib/internal/observable.js

```diff
@@ -0,0 +1,898 @@
+      return new Observable((subscriber) => {
+        value.then(
+          (value) => {
+            subscriber.next(value);
+            subscriber.complete();
+          },
+          (error) => subscriber.error(error)
+        );
+      });
+    }
+
+    if (typeof value?.[SymbolAsyncIterator] === 'function' ||
+        typeof value?.[SymbolIterator] === 'function') {
+      const processIterable = async (asyncIterable, subscriber) => {
+        if (!subscriber.active) return;
+        for await (const chunk of asyncIterable) {
+          if (!subscriber.active) return;
+          subscriber.next(chunk);
+        }
+        subscriber.complete();
+      };
+
+      return new Observable((subscriber) => {
+        processIterable(value, subscriber)
+          .catch((err) => subscriber.error(err));
+      });
+    }
+
+    throw new ERR_INVALID_ARG_TYPE('value',
+      ['Observable', 'Promise', 'AsyncIterable', 'Iterable'], value);
+  }
+
+  constructor(subscriberCallback) {
+    validateFunction(subscriberCallback, 'subscriberCallback');
+    this.#subscriberCallback = subscriberCallback;
+    this[kObserverMarker] = kObserverMarker;
+  }
+
+  /**
+   * @typedef {SubscriptionObserverCallback|SubscriptionObserver} ObserverUnion
+   *
+   * @param {ObserverUnion} observer
+   * @param {SubscribOptions} options
+   */
+  subscribe(observer = {}, options = {}) {
+    let internalObserver = {
+      next(value) {},
+      error(error) {
+        throw error;
+        // The spec expects us to use reportError();
+      },
+      complete() {},
+    };
+    if (typeof observer === 'function') {
+      internalObserver = {
+        next(value) {
+          try {
+            observer(value);
+          } catch (err) {
+            throw error;
+            // The spec expects us to use reportError();
+          }
+        }
+      };
+    } else if (typeof observer === 'object' && observer !== null) {
+      const { next, error, complete } = observer;
+      if (next != null) {
+        internalObserver.next = (value) => {
+          try {
+            next(value);
+          } catch (err) {
+            throw error;
+            // The spec expects us to use reportError();
+          }
+        };
+      }
+      if (error != null) {
+        internalObserver.error = (err) => {
+          try {
+            error(err);
+          } catch (err) {
+            throw error;
+            // The spec expects us to use reportError();
+          }
+        };
+      }
+      if (complete != null) {
+        internalObserver.complete = () => {
+          try {
+            complete();
+          } catch (err) {
+            throw error;
+            // The spec expects us to use reportError();
+          }
+        };
+      }
+    }
+    const subscriber = createSubscriber(internalObserver, options);
+    if (subscriber.signal.aborted) {
+      subscriber[kClose]();
+    } else {
+      subscriber.signal.addEventListener('abort', () => {
```

Comment:
This is a memory leak, please use the internal utility we have for subscribing to events without leaking. (i.e. pass `kWeakHandler: subscriber`)

Reply:
(or remove the event listener when the subscriber is closed)